File: java/exercises/inheritance/Progression.java
/** Generates a simple progression. By default 0, 1, 2,...*/
public class Progression {

	// instance variable
	protected long current;

	/** Constructs a progression starting at 0 */
	public Progression(){ this(0); }

	/** Construct a progression with given start */
	public Progression(long start) { current = start; }

	/** Returns the next value of the progression */
	public long nextValue(){
		long result = current;
		advance();
		return result;
	}

	/** Print the next n values of the progression, separated by space */
	public void printProgression(int n){
		System.out.print(nextValue());
		for(int i=1; i<n; i++){
			System.out.print(" " + nextValue());
		}
		System.out.println();
	}

	/** Advance the current value to the next value in the progression */
	protected void advance(){
		current++;
	}


}

File: python/exercises/inheritance/Progression.py
class Progression:
    """ Iterator producing a generic progression.

    Default iterator produces the whole numbers 0, 1, 2....
    """

    def __init__(self, start=0):
        """Initalize the first value of the progression."""
        self._current = start

    def _advance(self):
        """Update self._current to a new value.
        This should be overridden by the subclass.
        By convention, if current is set to None,
        this designates the end of a finite progression
        """
        self._current += 1

    def __next__(self):
        """Return the next element, or raise StopIteration error"""
        if self._current is None:
            raise StopIteration()
        else:
            result = self._current
            self._advance()
            return result

    def __iter__(self):
        """By convention, an iterator must return itself as an iterator"""
        return self

    def print_progression(self, n):
        """Print the next n values of the progression"""
        print(" ".join(str(next(self)) for _ in range(n)))
